vpc/service/assign_addresses_v3.go

```diff
@@ -98,27 +102,56 @@ type branchENI struct {
 	idx           int
 }
 
+func subnetCacheKey(az, accountID string, subnetIDs []string) string {
+	sort.Strings(subnetIDs)
+	return fmt.Sprintf("az:%s accountID:%s subnetIDs:%s", az, accountID, strings.Join(subnetIDs, ","))
+}
+
 func (vpcService *vpcService) getSubnet(ctx context.Context, az, accountID string, subnetIDs []string) (*subnet, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	ctx, span := trace.StartSpan(ctx, "getSubnet")
-	defer span.End()
+	// NOTE[jigish] caching using the full set of subnetIDs will cause the same subnet to be returned for all hits
+	cacheKey := subnetCacheKey(az, accountID, subnetIDs)
+	r, err, _ := vpcService.getSubnetLock.Do(cacheKey, func() (interface{}, error) {
+		ctx, cancel := context.WithCancel(ctx)
```

Comment:
I think you should have a span on the outside as well to indicate if multiple raced.

Reply:
added